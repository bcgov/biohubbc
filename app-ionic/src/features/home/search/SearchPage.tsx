import {
  Accordion,
  AccordionDetails,
  AccordionSummary,
  Box,
  Button,
  Container,
  makeStyles,
  Paper,
  Theme,
  Typography
} from '@material-ui/core';
import { ExpandMore } from '@material-ui/icons';
import FilterControlsContainer, { ISearchFilterCriteria } from 'components/search/FilterControlsContainer';
import PaginationControlsContainer from 'components/search/PaginationControlsContainer';
import { calculateTotalPages } from 'components/search/paginationUtils';
import SearchResultsList from 'components/search/SearchResultsList';
import { useBiohubApi } from 'hooks/useBiohubApi';
import { useQuery } from 'hooks/useQuery';
import { IActivitySearchCriteria, SORT_DIRECTION } from 'interfaces/useBiohubApi-interfaces';
import moment from 'moment';
import qs from 'qs';
import React, { useEffect, useState } from 'react';
import { useHistory } from 'react-router';

const useStyles = makeStyles((theme: Theme) => ({
  fieldset: {
    border: 'none',
    marginBottom: '1rem'
  }
}));

export interface ISearchActivity {
  _id: string;
  activityType: string;
  activitySubtype: string;
  dateReceived: string;
}

export interface ISearchPagePage {
  classes?: any;
}

const SearchPage: React.FC<ISearchPagePage> = (props) => {
  const classes = useStyles();

  const history = useHistory();

  const queryParams = useQuery();

  const biohubApi = useBiohubApi();

  const [searchFilterCriteria, setSearchFilterCriteria] = useState<ISearchFilterCriteria>({
    startDate: queryParams.startDate || '',
    endDate: queryParams.endDate || '',
    activityType: queryParams.activityType || [],
    speciesType: queryParams.speciesType || [],
    habitatType: queryParams.habitatType || []
  });

  const [currentPage, setCurrentPage] = useState(Number(queryParams.currentPage) || 1);
  const [itemsPerPage, setItemsPerPage] = useState(Number(queryParams.itemsPerPage) || 25);
  const [totalItems, setTotalItems] = useState(0);
  const [totalPages, setTotalPages] = useState(1);

  const [sortBy, setSortBy] = useState(queryParams.sortBy || '-received_timestamp');

  const [searchResults, setSearchResults] = useState<ISearchActivity[]>([]);

  const [isDisabled, setIsDisabled] = useState(false);

  const sortByValues = {
    // Note: prefixed with `+` or `-` to represent `ASC` or `DESC` respectively
    '-received_timestamp': 'Date (newest at top)',
    '+activity_type': 'Type (a-z)',
    '+activity_subtype': 'Subtype (a-z)'
  };

  /**
   * Call the API to fetch records based on the current filter criteria and pagination settings.
   */
  const handleSearch = async () => {
    setIsDisabled(true);

    // Save the `searchFilterCriteria` to the url
    history.push({ search: qs.stringify({ ...queryParams, ...searchFilterCriteria }) });

    const startDate = moment(searchFilterCriteria.startDate, 'YYYY-MM-DD');
    const startDateFilter: object =
      (startDate && startDate.isValid() && { date_range_start: startDate.startOf('day').toISOString() }) || {};

    const endDate = moment(searchFilterCriteria.endDate, 'YYYY-MM-DD');
    const endDateFilter: object =
      (endDate && endDate.isValid() && { date_range_end: endDate.endOf('day').toISOString() }) || {};

    const activityTypeFilter: object =
      (searchFilterCriteria.activityType &&
        searchFilterCriteria.activityType.length && { activity_type: searchFilterCriteria.activityType }) ||
      {};

    const sortColumn = sortBy.substring(1);
    const sortDirection = sortBy.substring(0, 1);

    const activitySearchCriteria: IActivitySearchCriteria = {
      page: currentPage - 1, // The api starts at page 0, while the UI starts at page 1.
      limit: itemsPerPage,
      sort_by: sortColumn,
      sort_direction: (sortDirection === '-' && SORT_DIRECTION.DESC) || SORT_DIRECTION.ASC,
      column_names: ['activity_id', 'activity_type', 'activity_subtype', 'created_timestamp', 'received_timestamp'],
      ...startDateFilter,
      ...endDateFilter,
      ...activityTypeFilter // NICK TODO WIP - Update backend search to accept array of type, etc. Move it all into the URL????
    };

    try {
      const response = await biohubApi.getActivities(activitySearchCriteria); // NICK TODO WIP - Need a simpler api that just fetches basic information as well as counts, and not photos, etc.

      if (!response || !response.rows) {
        // TODO error messaging
      }

      const activities = response.rows.map((activity) => {
        return {
          _id: activity.activity_id,
          activityId: activity.activity_id,
          activityType: activity.activity_type,
          activitySubtype: activity.activity_subtype,
          dateCreated: activity.created_timestamp,
          dateReceived: activity.received_timestamp
        };
      });

      const count = Number(response.count);

      setSearchResults(activities);
      setTotalItems(count);
      setTotalPages(calculateTotalPages(count, itemsPerPage));
    } catch {
      // TODO error messaging
    }

    setIsDisabled(false);
  };

  useEffect(() => {
    // Save the new `itemsPerPage` to the url
    history.push({ search: qs.stringify({ ...queryParams, itemsPerPage: itemsPerPage }) });

    if (currentPage !== 1) {
      // The current page needs to be reset on `itemsPerPage` change
      // In this case, do not trigger the search, as the useEffect on [currentPage] will trigger the search
      setCurrentPage(1);
      return;
    }

    handleSearch();
  }, [itemsPerPage]);

  useEffect(() => {
    // Save the new `currentPage` to the url
    history.push({ search: qs.stringify({ ...queryParams, currentPage: currentPage }) });

    handleSearch();
  }, [currentPage]);

  useEffect(() => {
    // Save the new `sortBy` to the url
    history.push({ search: qs.stringify({ ...queryParams, sortBy: sortBy }) });

    handleSearch();
  }, [sortBy]);

  return (
    <Container className={props.classes.container}>
      <Box mb={3}>
        <Typography variant="h4">Search biohubbc Records</Typography>
      </Box>
      <Box mb={2}>
        <form className={props.classes.form}>
          <Accordion defaultExpanded={true}>
            <AccordionSummary
              expandIcon={<ExpandMore />}
              aria-controls="search-filter-controls"
              id="search-filter-controls">
              <Typography variant="h5">Filters</Typography>
            </AccordionSummary>
            <AccordionDetails>
              <Box flexDirection="column">
                <Box>
                  <FilterControlsContainer
                    classes={classes}
                    searchFilterCriteriaState={{ searchFilterCriteria, setSearchFilterCriteria }}
                    isDisabled={isDisabled}
                  />
                </Box>
                <Box p={2}>
                  <Button variant="contained" color="primary" onClick={handleSearch} disabled={isDisabled}>
                    Search
                  </Button>
                </Box>
              </Box>
            </AccordionDetails>
          </Accordion>
        </form>
      </Box>
      <Paper>
        <Box p={2}>
          <Box>
            <PaginationControlsContainer
              classes={classes}
              currentPageState={{ currentPage, setCurrentPage }}
              itemsPerPageState={{ itemsPerPage, setItemsPerPage }}
              totalItems={totalItems}
              totalPages={totalPages}
              sortByValues={sortByValues}
              sortByState={{ sortBy, setSortBy }}
              isDisabled={isDisabled}
            />
          </Box>
          <Box>
            <SearchResultsList classes={classes} totalItems={totalItems} results={searchResults} />
          </Box>
        </Box>
      </Paper>
    </Container>
  );
};

export default SearchPage;
