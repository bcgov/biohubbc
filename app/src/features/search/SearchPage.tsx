import Box from '@material-ui/core/Box';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import centerOfMass from '@turf/center-of-mass';
import { IErrorDialogProps } from 'components/dialog/ErrorDialog';
import MapContainer, { IClusteredPointGeometries } from 'components/map/MapContainer';
import { SearchFeaturePopup } from 'components/map/SearchFeaturePopup';
import { DialogContext } from 'contexts/dialogContext';
import { APIError } from 'hooks/api/useAxios';
import { useBiohubApi } from 'hooks/useBioHubApi';
import React, { useCallback, useContext, useEffect, useState } from 'react';
import { generateValidGeometryCollection } from 'utils/mapBoundaryUploadHelpers';

/**
 * Page to search for and display a list of records spatially.
 *
 * @return {*}
 */
const SearchPage: React.FC = () => {
  const biohubApi = useBiohubApi();

  const [performSearch, setPerformSearch] = useState<boolean>(true);
  const [geometries, setGeometries] = useState<IClusteredPointGeometries[]>([]);

  const dialogContext = useContext(DialogContext);

  const showFilterErrorDialog = useCallback(
    (textDialogProps?: Partial<IErrorDialogProps>) => {
      dialogContext.showErrorDialog({
        onClose: () => {
          dialogContext.hideDialog();
        },
        onOk: () => {
          dialogContext.hideDialog();
        },
        ...textDialogProps
      });
    },
    [dialogContext]
  );

  const getSearchResults = useCallback(async () => {
    try {
      const response = await biohubApi.search.getSearchResults();

      if (!response) {
        setPerformSearch(false);
        return;
      }

      const clusteredPointGeometries: IClusteredPointGeometries[] = [];

      response.forEach((result: any) => {
        const feature = generateValidGeometryCollection(result.geometry, result.id).geometryCollection[0];

        clusteredPointGeometries.push({
          coordinates: centerOfMass(feature as any).geometry.coordinates,
          popupComponent: <SearchFeaturePopup featureData={result} />
        });
      });

      setPerformSearch(false);
      setGeometries(clusteredPointGeometries);
    } catch (error) {
      const apiError = error as APIError;
      showFilterErrorDialog({
        dialogTitle: 'Error Searching For Results',
        dialogError: apiError?.message,
        dialogErrorDetails: apiError?.errors
      });
    }
  }, [biohubApi.search, showFilterErrorDialog]);

  useEffect(() => {
    if (performSearch) {
      getSearchResults();
    }
  }, [performSearch, getSearchResults]);

  /**
   * Displays search results visualized on a map spatially.
   */
  return (
    <Box my={4}>
      <Container maxWidth="xl">
        <Box mb={5} display="flex" justifyContent="space-between">
          <Typography variant="h1">Map</Typography>
        </Box>
        <Box>
          <Box mb={4}>
            <Grid item xs={12}>
              <Box mt={2} height={750}>
                <MapContainer
                  mapId="search_boundary_map"
                  scrollWheelZoom={true}
                  clusteredPointGeometries={geometries}
                />
              </Box>
            </Grid>
          </Box>
        </Box>
      </Container>
    </Box>
  );
};

export default SearchPage;
