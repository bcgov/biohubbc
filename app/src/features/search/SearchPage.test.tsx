import { render } from '@testing-library/react';
import React from 'react';
import SearchPage from './SearchPage';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

const history = createMemoryHistory();

jest.spyOn(console, 'debug').mockImplementation(() => {});

describe('SearchPage', () => {
  it('renders correctly', () => {
    const { getByText } = render(
      <Router history={history}>
        <SearchPage />
      </Router>
    );

    expect(getByText('Map')).toBeInTheDocument();
  });
});
