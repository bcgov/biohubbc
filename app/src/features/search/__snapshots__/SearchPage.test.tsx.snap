// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-1"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthXl"
    >
      <div
        class="MuiBox-root MuiBox-root-2"
      >
        <h1
          class="MuiTypography-root MuiTypography-h1"
        >
          Search
        </h1>
      </div>
      <div
        class="MuiBox-root MuiBox-root-3"
      >
        <div
          class="MuiBox-root MuiBox-root-4"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiBox-root MuiBox-root-5"
            >
              <div
                class="leaflet-container leaflet-grab leaflet-touch-drag"
                id="search_boundary_map"
                style="height: 100%; position: relative;"
                tabindex="0"
              >
                <div
                  class="leaflet-pane leaflet-map-pane"
                  style="left: 0px; top: 0px;"
                >
                  <div
                    class="leaflet-pane leaflet-tile-pane"
                  >
                    <div
                      class="leaflet-layer "
                      style="z-index: 1;"
                    >
                      <div
                        class="leaflet-tile-container leaflet-zoom-animated"
                        style="z-index: 18; left: 0px; top: 0px;"
                      >
                        <img
                          alt=""
                          class="leaflet-tile"
                          role="presentation"
                          src="https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/5/10/4"
                          style="width: 256px; height: 256px; left: -159px; top: -31px;"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="leaflet-pane leaflet-shadow-pane"
                  />
                  <div
                    class="leaflet-pane leaflet-overlay-pane"
                  >
                    <svg
                      height="0"
                      pointer-events="none"
                      style="left: 0px; top: 0px;"
                      viewBox="0 0 0 0"
                      width="0"
                    >
                      <g />
                    </svg>
                  </div>
                  <div
                    class="leaflet-pane leaflet-marker-pane"
                  />
                  <div
                    class="leaflet-pane leaflet-tooltip-pane"
                  />
                  <div
                    class="leaflet-pane leaflet-popup-pane"
                  />
                </div>
                <div
                  class="leaflet-control-container"
                >
                  <div
                    class="leaflet-top leaflet-left"
                  >
                    <div
                      class="leaflet-control-zoom leaflet-bar leaflet-control"
                    >
                      <a
                        aria-label="Zoom in"
                        class="leaflet-control-zoom-in"
                        href="#"
                        role="button"
                        title="Zoom in"
                      >
                        +
                      </a>
                      <a
                        aria-label="Zoom out"
                        class="leaflet-control-zoom-out"
                        href="#"
                        role="button"
                        title="Zoom out"
                      >
                        âˆ’
                      </a>
                    </div>
                  </div>
                  <div
                    class="leaflet-top leaflet-right"
                  >
                    <div
                      class="leaflet-draw leaflet-control"
                    >
                      <div
                        class="leaflet-draw-section"
                      >
                        <div
                          class="leaflet-draw-toolbar leaflet-bar leaflet-draw-toolbar-top"
                        >
                          <a
                            class="leaflet-draw-draw-polyline"
                            href="#"
                            title="Draw a polyline"
                          >
                            <span
                              class="sr-only"
                            >
                              Draw a polyline
                            </span>
                          </a>
                          <a
                            class="leaflet-draw-draw-polygon"
                            href="#"
                            title="Draw a polygon"
                          >
                            <span
                              class="sr-only"
                            >
                              Draw a polygon
                            </span>
                          </a>
                          <a
                            class="leaflet-draw-draw-rectangle"
                            href="#"
                            title="Draw a rectangle"
                          >
                            <span
                              class="sr-only"
                            >
                              Draw a rectangle
                            </span>
                          </a>
                          <a
                            class="leaflet-draw-draw-marker"
                            href="#"
                            title="Draw a marker"
                          >
                            <span
                              class="sr-only"
                            >
                              Draw a marker
                            </span>
                          </a>
                          <a
                            class="leaflet-draw-draw-circlemarker"
                            href="#"
                            title="Draw a circlemarker"
                          >
                            <span
                              class="sr-only"
                            >
                              Draw a circlemarker
                            </span>
                          </a>
                        </div>
                        <ul
                          class="leaflet-draw-actions"
                        />
                      </div>
                      <div
                        class="leaflet-draw-section"
                      >
                        <div
                          class="leaflet-draw-toolbar leaflet-bar"
                        >
                          <a
                            class="leaflet-draw-edit-edit leaflet-disabled"
                            href="#"
                            title="No layers to edit"
                          >
                            <span
                              class="sr-only"
                            >
                              Edit layers
                            </span>
                          </a>
                          <a
                            class="leaflet-draw-edit-remove leaflet-disabled"
                            href="#"
                            title="No layers to delete"
                          >
                            <span
                              class="sr-only"
                            >
                              Delete layers
                            </span>
                          </a>
                        </div>
                        <ul
                          class="leaflet-draw-actions"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="leaflet-bottom leaflet-left"
                  />
                  <div
                    class="leaflet-bottom leaflet-right"
                  >
                    <div
                      aria-haspopup="true"
                      class="leaflet-control-layers leaflet-control"
                    >
                      <a
                        class="leaflet-control-layers-toggle"
                        href="#"
                        title="Layers"
                      />
                      <section
                        class="leaflet-control-layers-list"
                      >
                        <div
                          class="leaflet-control-layers-base"
                        >
                          <label>
                            <div>
                              <input
                                checked="checked"
                                class="leaflet-control-layers-selector"
                                name="leaflet-base-layers_49"
                                type="radio"
                              />
                              <span>
                                 Esri Imagery
                              </span>
                            </div>
                          </label>
                          <label>
                            <div>
                              <input
                                class="leaflet-control-layers-selector"
                                name="leaflet-base-layers_49"
                                type="radio"
                              />
                              <span>
                                 BC Government
                              </span>
                            </div>
                          </label>
                        </div>
                        <div
                          class="leaflet-control-layers-separator"
                        />
                        <div
                          class="leaflet-control-layers-overlays"
                        >
                          <label>
                            <div>
                              <input
                                class="leaflet-control-layers-selector"
                                type="checkbox"
                              />
                              <span>
                                 Wildlife Management Units
                              </span>
                            </div>
                          </label>
                          <label>
                            <div>
                              <input
                                class="leaflet-control-layers-selector"
                                type="checkbox"
                              />
                              <span>
                                 Parks - Section
                              </span>
                            </div>
                          </label>
                          <label>
                            <div>
                              <input
                                class="leaflet-control-layers-selector"
                                type="checkbox"
                              />
                              <span>
                                 Parks - Regional
                              </span>
                            </div>
                          </label>
                        </div>
                      </section>
                    </div>
                    <div
                      class="leaflet-control-attribution leaflet-control"
                    >
                      <a
                        href="https://leafletjs.com"
                        title="A JS library for interactive maps"
                      >
                        Leaflet
                      </a>
                       | Â© 
                      <a
                        href="https://www.esri.com/en-us/arcgis/products/location-services/services/basemaps"
                      >
                        ESRI Basemap
                      </a>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
